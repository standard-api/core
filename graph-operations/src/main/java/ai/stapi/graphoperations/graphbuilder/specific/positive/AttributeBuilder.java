package ai.stapi.graphoperations.graphbuilder.specific.positive;

import ai.stapi.graph.attribute.Attribute;
import ai.stapi.graph.attribute.attributeFactory.AttributeValueFactoryInput;
import ai.stapi.graph.attribute.attributeFactory.GenericAttributeFactory;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class AttributeBuilder {

  private String attributeName;
  private String structureType;
  private List<AttributeValueFactoryInput> attributeValues;

  public AttributeBuilder() {
    this.attributeValues = new ArrayList<>();
  }

  public AttributeBuilder setAttributeName(String attributeName) {
    this.attributeName = attributeName;
    return this;
  }

  public AttributeBuilder addAttributeValue(AttributeValueFactoryInput attributeValue) {
    this.attributeValues.add(attributeValue);
    return this;
  }


  public AttributeBuilder setAttributeStructureType(String structureType) {
    this.structureType = structureType;
    return this;
  }

  public boolean isComplete() {
    return this.attributeName != null && this.attributeValues != null
        && this.structureType != null;
  }

  public Attribute<?> build(GenericAttributeFactory attributeFactory) {
    return attributeFactory.create(
        this.attributeName,
        this.structureType,
        this.attributeValues,
        new HashMap<>()
    );
  }
}
