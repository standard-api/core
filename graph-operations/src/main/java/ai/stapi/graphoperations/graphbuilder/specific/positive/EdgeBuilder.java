package ai.stapi.graphoperations.graphbuilder.specific.positive;

public interface EdgeBuilder extends GraphElementBuilder {

  EdgeDirection getEdgeDirection();

  EdgeBuilder setEdgeDirection(EdgeDirection edgeDirection);
}
