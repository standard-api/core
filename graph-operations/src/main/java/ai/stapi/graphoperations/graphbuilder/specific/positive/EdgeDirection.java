package ai.stapi.graphoperations.graphbuilder.specific.positive;

public enum EdgeDirection {
  INGOING,
  OUTGOING
}
