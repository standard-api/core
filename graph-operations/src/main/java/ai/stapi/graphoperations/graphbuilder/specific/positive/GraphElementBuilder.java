package ai.stapi.graphoperations.graphbuilder.specific.positive;

import ai.stapi.graph.attribute.attributeFactory.GenericAttributeFactory;
import ai.stapi.graph.Graph;
import ai.stapi.graph.graphelements.GraphElement;
import ai.stapi.identity.UniqueIdentifier;

public interface GraphElementBuilder {

  GraphElementBuilder setId(UniqueIdentifier id);

  GraphElementBuilder setType(String type);

  AttributeBuilder addAttribute();

  AttributeBuilder getLastAttribute();

  boolean isComplete();

  GraphElement build(GenericAttributeFactory attributeFactory);

  Graph buildToGraph(GenericAttributeFactory attributeFactory);
}
