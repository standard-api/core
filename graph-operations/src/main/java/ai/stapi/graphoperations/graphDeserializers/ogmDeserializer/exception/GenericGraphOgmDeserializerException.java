package ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exception;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.PositiveGraphDescription;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.graphoperations.ogmProviders.specific.objectGraphMappingMappingProviders.OgmGraphElementTypes;
import ai.stapi.graphoperations.graphReader.exception.GraphReaderException;
import ai.stapi.graphoperations.graphReader.readResults.ReadResult;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectObjectGraphMapping;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.graphoperations.serializationTypeProvider.specific.SpecificSerializationTypeProvider;
import java.util.List;
import org.apache.commons.lang3.StringUtils;

public class GenericGraphOgmDeserializerException extends RuntimeException {

  private GenericGraphOgmDeserializerException(String message) {
    super(message);
  }

  public static GenericGraphOgmDeserializerException becauseNodeTypeIsNotSupported(
      String nodeType) {
    return new GenericGraphOgmDeserializerException(
        "There are no supporting "
            + SpecificSerializationTypeProvider.class.getSimpleName()
            + " provider for node of type '"
            + nodeType + "'."
    );
  }

  public static GenericGraphOgmDeserializerException becauseNodeTypeIsSupportedByMultipleDeserializers(
      String nodeType,
      List<String> supportingDeserializerClassNames
  ) {
    return new GenericGraphOgmDeserializerException(
        "Node of type '" + nodeType
            + "' is supported by multiple deserializers: ["
            + StringUtils.join(supportingDeserializerClassNames, ", ")
            + "]."
    );
  }

  public static GenericGraphOgmDeserializerException becauseDeserializationHasToStartFromObjectGraphMapping(
      String providedNodeType) {
    return new GenericGraphOgmDeserializerException(
        "Deserialization has to start from "
            + ObjectObjectGraphMapping.class.getSimpleName()
            + " serialized under node '"
            + OgmGraphElementTypes.OGM_OBJECT_NODE
            + "'. Provided node type was: '"
            + providedNodeType + "'."
    );
  }

  public static GenericGraphOgmDeserializerException becauseThereIsMultipleGraphBranchesButMappingIsNotForList() {
    return new GenericGraphOgmDeserializerException(
        "Graph description led to multiple graph branches but provided "
            + ObjectGraphMapping.class.getSimpleName()
            + " did not point to list."
    );
  }

  public static GenericGraphOgmDeserializerException becauseGraphTraversingLedToMultipleBranches(
      String elementType) {
    return new GenericGraphOgmDeserializerException(
        GraphDescription.class.getSimpleName()
            + " inside " + ObjectObjectGraphMapping.class.getSimpleName()
            + " led from element of type '" + elementType
            + "' to multiple graph branches and that is not allowed."
    );
  }

  public static GenericGraphOgmDeserializerException becauseGraphReadingAccordingToOgmFailed(
      String currentElementType,
      PositiveGraphDescription declaration,
      GraphReaderException exception
  ) {
    return new GenericGraphOgmDeserializerException(
        "Traversing graph from element of type '" + currentElementType
            + "' using " + declaration.getClass().getSimpleName()
            + " failed. Ogm doesn't correspond to the actual graph structure saved. "
            + System.lineSeparator()
            + "Cause: " + exception.getMessage()
    );
  }

  public static GenericGraphOgmDeserializerException becauseElementWasNotFound(
      UniqueIdentifier graphElementId) {
    return new GenericGraphOgmDeserializerException(
        "Graph element with id [" + graphElementId
            + "] was not found in graph. This should have never happen!"
    );
  }

  public static GenericGraphOgmDeserializerException givenReadResultIsNotGraphElement() {
    return new GenericGraphOgmDeserializerException(
        ReadResult.class.getSimpleName() + " does not contain traversable graph element in it."
    );
  }

  public static GenericGraphOgmDeserializerException becauseGraphTraversingLedToNoBranches(
      String type) {
    return new GenericGraphOgmDeserializerException(
        GraphDescription.class.getSimpleName()
            + " inside " + ObjectObjectGraphMapping.class.getSimpleName()
            + " led from element of type '" + type
            + "' to no graph branches and that is not allowed."
    );
  }
}
