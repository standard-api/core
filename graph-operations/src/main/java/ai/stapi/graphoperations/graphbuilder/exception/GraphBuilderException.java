package ai.stapi.graphoperations.graphbuilder.exception;

import ai.stapi.graphoperations.graphbuilder.specific.positive.NodeBuilder;
import ai.stapi.graphoperations.graphbuilder.specific.positive.GraphElementBuilder;

public class GraphBuilderException extends RuntimeException {

  private GraphBuilderException(String message) {
    super(message);
  }

  public static GraphBuilderException becauseAttributeValueTypeIsNotSupported(Object value) {
    return new GraphBuilderException(
        "Attribute value of type '" + value.toString() + "' is not supported.");
  }

  public static GraphBuilderException becauseAttributeBuilderIsNotComplete() {
    return new GraphBuilderException("Attribute builder is not complete.");
  }

  public static GraphBuilderException becauseNodeIsMissingIdOrType() {
    return new GraphBuilderException("Node is missing id or type.");
  }

  public static GraphBuilderException becauseEdgeIsNotCompleted() {
    return new GraphBuilderException(
        "Edge was not complete. It is missing either id, type or direction.");
  }

  public static GraphBuilderException becauseNodeFromOnEdgeIsNotComplete() {
    return new GraphBuilderException("Node FROM on edge is not complete.");
  }

  public static GraphBuilderException becauseNodeToOnEdgeIsNotComplete() {
    return new GraphBuilderException("Node TO on edge is not complete.");
  }

  public static GraphBuilderException becauseGraphCanNotHaveEdgeAsFirstElement() {
    return new GraphBuilderException("Graph can not have edge as first element.");
  }

  public static GraphBuilderException becauseGraphCanNotHaveEdgeAsLastElement() {
    return new GraphBuilderException("Graph can not have edge as last element.");
  }

  public static GraphBuilderException becauseOfUnexpectedEdgeNeighbor(
      GraphElementBuilder elementBuilder) {
    return new GraphBuilderException(
        "Because there was unexpected edge neighbor: '"
            + (elementBuilder == null ? "null'."
            : elementBuilder.getClass().getSimpleName() + "'."));
  }

  public static GraphBuilderException becauseThereCanNotBeTwoNodesAfterEachOther(
      NodeBuilder previousNode) {
    return new GraphBuilderException(
        "There can not be two nodes after each other. Previous node type: '"
            + previousNode.getType() + "'."
    );
  }

  public static GraphBuilderException becauseThereCanNotBeTwoEdgesAfterEachOther() {
    return new GraphBuilderException("There can not be two edges after each other.");
  }

  public static GraphBuilderException becauseEdgeHasToBePrecededByNode() {
    return new GraphBuilderException("Edge has to be preceded by node.");
  }

  public static GraphBuilderException becauseThereAreNoElementsAbleToContainAttribute() {
    return new GraphBuilderException("There are no elements able to contain attributes.");
  }

  public static GraphBuilderException becauseThereAreNoAttributesOnElement() {
    return new GraphBuilderException("There are no attributes on element.");
  }

  public static GraphBuilderException becauseThereAreNoElements() {
    return new GraphBuilderException("There are no elements inside builder.");
  }
}
