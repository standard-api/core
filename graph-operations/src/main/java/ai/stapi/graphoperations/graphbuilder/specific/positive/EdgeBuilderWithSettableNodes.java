package ai.stapi.graphoperations.graphbuilder.specific.positive;

import ai.stapi.graphoperations.graphbuilder.exception.GraphBuilderException;
import ai.stapi.graph.attribute.Attribute;
import ai.stapi.graph.attribute.attributeFactory.GenericAttributeFactory;
import ai.stapi.graph.Graph;
import ai.stapi.graph.graphelements.Edge;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.identity.UniqueIdentifier;
import java.util.ArrayList;
import java.util.List;

public class EdgeBuilderWithSettableNodes implements EdgeBuilder {

  private final List<AttributeBuilder> attributes = new ArrayList<>();
  private NodeBuilder precedingNode;
  private NodeBuilder nodeFrom;
  private NodeBuilder nodeTo;
  private EdgeDirection edgeDirection;
  private UniqueIdentifier id;
  private String type;

  @Override
  public EdgeBuilder setId(UniqueIdentifier id) {
    this.id = id;
    return this;
  }

  @Override
  public EdgeBuilder setType(String type) {
    this.type = type;
    return this;
  }

  @Override
  public AttributeBuilder addAttribute() {
    var attributeBuilder = new AttributeBuilder();
    attributes.add(attributeBuilder);
    return attributeBuilder;
  }

  @Override
  public AttributeBuilder getLastAttribute() {
    if (this.attributes.size() == 0) {
      throw GraphBuilderException.becauseThereAreNoAttributesOnElement();
    }
    return this.attributes.get(attributes.size() - 1);
  }

  @Override
  public boolean isComplete() {
    return this.id != null
        && this.type != null
        && this.edgeDirection != null
        && this.nodeTo != null
        && this.nodeFrom != null
        && this.nodeFrom.isComplete()
        && this.nodeTo.isComplete();
  }

  @Override
  public Edge build(GenericAttributeFactory attributeFactory) {
    if (this.nodeTo == null) {
      throw GraphBuilderException.becauseGraphCanNotHaveEdgeAsLastElement();
    }
    if (!this.isComplete()) {
      throw GraphBuilderException.becauseEdgeIsNotCompleted();
    }
    if (!this.nodeFrom.isComplete()) {
      throw GraphBuilderException.becauseNodeFromOnEdgeIsNotComplete();
    }
    if (!this.nodeTo.isComplete()) {
      throw GraphBuilderException.becauseNodeToOnEdgeIsNotComplete();
    }
    var nodeFrom = this.nodeFrom.build(attributeFactory);
    var nodeTo = this.nodeTo.build(attributeFactory);
    var edge = new Edge(this.id, nodeFrom, this.type, nodeTo);
    for (AttributeBuilder attributeBuilder : this.attributes) {
      Attribute<?> build = attributeBuilder.build(attributeFactory);
      edge = edge.add(build);
    }
    return edge;
  }

  @Override
  public Graph buildToGraph(GenericAttributeFactory attributeFactory) {
    Edge edge = this.build(attributeFactory);
    Node nodeFrom = this.nodeFrom.build(attributeFactory);
    Node nodeTo = this.nodeTo.build(attributeFactory);
    return new Graph(
        nodeFrom,
        nodeTo,
        edge
    );
  }

  public EdgeBuilderWithSettableNodes setPrecedingNode(NodeBuilder nodeBuilder) {
    this.precedingNode = nodeBuilder;
    return this;
  }

  public EdgeBuilderWithSettableNodes setNodeFrom(NodeBuilder nodeFrom) {
    this.nodeFrom = nodeFrom;
    return this;
  }

  public EdgeBuilderWithSettableNodes setNodeTo(NodeBuilder nodeTo) {
    this.nodeTo = nodeTo;
    return this;
  }

  @Override
  public EdgeBuilder setEdgeDirection(EdgeDirection edgeDirection) {
    if (precedingNode == null) {
      throw GraphBuilderException.becauseEdgeHasToBePrecededByNode();
    }
    this.edgeDirection = edgeDirection;
    if (edgeDirection.equals(EdgeDirection.INGOING)) {
      this.nodeTo = this.precedingNode;
    } else if (edgeDirection.equals(EdgeDirection.OUTGOING)) {
      this.nodeFrom = this.precedingNode;
    }
    return this;
  }

  @Override
  public EdgeDirection getEdgeDirection() {
    return edgeDirection;
  }
}
