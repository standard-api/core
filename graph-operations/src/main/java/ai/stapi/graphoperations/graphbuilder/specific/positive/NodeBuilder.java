package ai.stapi.graphoperations.graphbuilder.specific.positive;

import ai.stapi.graphoperations.graphbuilder.exception.GraphBuilderException;
import ai.stapi.graph.attribute.Attribute;
import ai.stapi.graph.attribute.attributeFactory.GenericAttributeFactory;
import ai.stapi.graph.Graph;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.identity.UniqueIdentifier;
import java.util.ArrayList;
import java.util.List;

public class NodeBuilder implements GraphElementBuilder {

  private final List<AttributeBuilder> attributes = new ArrayList<>();
  private UniqueIdentifier id;
  private String type;

  @Override
  public NodeBuilder setId(UniqueIdentifier id) {
    this.id = id;
    return this;
  }

  public NodeBuilder setId(String id) {
    this.id = new UniqueIdentifier(id);
    return this;
  }

  @Override
  public AttributeBuilder addAttribute() {
    var attributeBuilder = new AttributeBuilder();
    attributes.add(attributeBuilder);
    return attributeBuilder;
  }

  @Override
  public AttributeBuilder getLastAttribute() {
    if (this.attributes.size() == 0) {
      throw GraphBuilderException.becauseThereAreNoAttributesOnElement();
    }
    return this.attributes.get(attributes.size() - 1);
  }

  @Override
  public boolean isComplete() {
    return this.id != null && this.type != null;
  }

  @Override
  public Node build(GenericAttributeFactory attributeFactory) {
    if (!this.isComplete()) {
      throw GraphBuilderException.becauseNodeIsMissingIdOrType();
    }
    var node = new Node(this.id, this.type);
    for (AttributeBuilder builder : this.attributes) {
      Attribute<?> attribute = builder.build(attributeFactory);
      node = node.add(attribute);
    }
    return node;
  }

  @Override
  public Graph buildToGraph(GenericAttributeFactory attributeFactory) {
    return new Graph(this.build(attributeFactory));
  }

  public String getType() {
    return type;
  }

  @Override
  public NodeBuilder setType(String type) {
    this.type = type;
    return this;
  }
}
