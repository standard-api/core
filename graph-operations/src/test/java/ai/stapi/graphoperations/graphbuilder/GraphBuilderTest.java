package ai.stapi.graphoperations.graphbuilder;

import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeFactory.AttributeValueFactoryInput;
import ai.stapi.graph.attribute.attributeFactory.GenericAttributeFactory;
import ai.stapi.graph.attribute.attributeValue.BooleanAttributeValue;
import ai.stapi.graph.attribute.attributeValue.DecimalAttributeValue;
import ai.stapi.graph.attribute.attributeValue.IntegerAttributeValue;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graphoperations.graphbuilder.exception.GraphBuilderException;
import ai.stapi.graphoperations.graphbuilder.specific.positive.EdgeDirection;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.objectRenderer.infrastructure.objectToJsonStringRenderer.ObjectToJSonStringOptions;
import ai.stapi.test.integration.IntegrationTestCase;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

public class GraphBuilderTest extends IntegrationTestCase {

  @Autowired
  private GenericAttributeFactory genericAttributeFactory;

  @Test
  public void itThrowsErrorWhenNodeIsMissingType() {
    var builder = new GraphBuilder();
    builder.addNode().setId(UniversallyUniqueIdentifier.randomUUID());
    Executable throwable = () -> builder.build(this.genericAttributeFactory);
    this.thenExceptionMessageApproved(GraphBuilderException.class, throwable);
  }

  @Test
  public void itThrowsErrorWhenEdgeIsFirst() {
    var builder = new GraphBuilder();
    Executable throwable = () -> builder.addEdge().setId(UniversallyUniqueIdentifier.randomUUID())
        .setType("irrelevant");
    this.thenExceptionMessageApproved(GraphBuilderException.class, throwable);
  }

  @Test
  public void itThrowsErrorWhenEdgeIsLast() {
    var builder = new GraphBuilder();
    builder.addNode().setId(UniversallyUniqueIdentifier.randomUUID()).setType("irrelevant");
    builder.addEdge().setEdgeDirection(EdgeDirection.OUTGOING)
        .setId(UniversallyUniqueIdentifier.randomUUID()).setType("irrelevant");
    Executable throwable = () -> builder.build(this.genericAttributeFactory);
    this.thenExceptionMessageApproved(GraphBuilderException.class, throwable);
  }

  @Test
  public void itThrowsErrorWhenNodeIsMissingId() {
    var builder = new GraphBuilder();
    builder.addNode().setType("example_type");
    Executable throwable = () -> builder.build(this.genericAttributeFactory);
    this.thenExceptionMessageApproved(GraphBuilderException.class, throwable);
  }

  @Test
  public void itCanCreateNodeWithSpecificIdAndType() {
    var id = UniversallyUniqueIdentifier.fromString("b7842bbd-09c5-4ffe-946f-38b80d1f42ad");
    var nodeType = "example_type";
    var builder = new GraphBuilder();
    builder.addNode()
        .setId(id)
        .setType(nodeType);
    var graph = builder.build(this.genericAttributeFactory);
    graph.getNode(id, nodeType);
    this.thenGraphApproved(graph);
  }

  @Test
  public void itCanCreateEdgeWithSpecificIdAndType() {
    var id = UniversallyUniqueIdentifier.fromString("2a9f9fdc-7f6b-4cc7-bd86-daa6deaa6152");
    var builder = new GraphBuilder();
    builder.addNode().setId(UniversallyUniqueIdentifier.randomUUID()).setType("irrelevant_type");
    builder.addEdge().setEdgeDirection(EdgeDirection.OUTGOING).setId(id).setType("example_edge");
    builder.addNode().setId(UniversallyUniqueIdentifier.randomUUID())
        .setType("another_irrelevant_type");
    var graph = builder.build(this.genericAttributeFactory);
    graph.getEdge(id, "example_edge");
    this.thenGraphApproved(graph);
  }

  @Test
  public void itCanCreateMultipleEdgesWithSpecificIdAndType() {
    var firstId = UniversallyUniqueIdentifier.fromString("2a9f9fdc-7f6b-4cc7-bd86-daa6deaa6152");
    var secondId = UniversallyUniqueIdentifier.fromString("2d3897d1-baf4-4578-a45b-bb8c07267b6e");
    var builder = new GraphBuilder();
    builder.addNode().setId(UniversallyUniqueIdentifier.randomUUID()).setType("first_node");
    builder.addEdge().setEdgeDirection(EdgeDirection.OUTGOING).setId(firstId).setType("first_edge");
    builder.addNode().setId(UniversallyUniqueIdentifier.randomUUID()).setType("second_node");
    builder.addEdge().setEdgeDirection(EdgeDirection.INGOING).setId(secondId)
        .setType("second_edge");
    builder.addNode().setId(UniversallyUniqueIdentifier.randomUUID()).setType("third_node");
    var graph = builder.build(this.genericAttributeFactory);
    graph.getEdge(firstId, "first_edge");
    graph.getEdge(secondId, "second_edge");
    this.thenGraphApproved(graph);
  }

  @Test
  public void itCanCreateNodeWithSpecificIdAndTypeAndAttribute() {
    var id = UniversallyUniqueIdentifier.fromString("b7842bbd-09c5-4ffe-946f-38b80d1f42ad");
    var nodeType = "example_type";
    var builder = new GraphBuilder();
    var nodeBuilder = builder.addNode()
        .setId(id)
        .setType(nodeType);
    nodeBuilder.addAttribute()
        .setAttributeName("string_attribute")
        .addAttributeValue(
            new AttributeValueFactoryInput(
                "string_value",
                StringAttributeValue.SERIALIZATION_TYPE
            )
        )
        .setAttributeStructureType(LeafAttribute.DATA_STRUCTURE_TYPE);
    nodeBuilder.addAttribute()
        .setAttributeName("integer_attribute")
        .addAttributeValue(
            new AttributeValueFactoryInput(
                1,
                IntegerAttributeValue.SERIALIZATION_TYPE
            )
        )
        .setAttributeStructureType(LeafAttribute.DATA_STRUCTURE_TYPE);
    nodeBuilder.addAttribute()
        .setAttributeName("boolean_attribute")
        .addAttributeValue(
            new AttributeValueFactoryInput(
                true,
                BooleanAttributeValue.SERIALIZATION_TYPE
            )
        )
        .setAttributeStructureType(LeafAttribute.DATA_STRUCTURE_TYPE);
    nodeBuilder.addAttribute()
        .setAttributeName("decimal_attribute")
        .addAttributeValue(
            new AttributeValueFactoryInput(
                5.6231,
                DecimalAttributeValue.SERIALIZATION_TYPE
            )
        )
        .setAttributeStructureType(LeafAttribute.DATA_STRUCTURE_TYPE);

    var graph = builder.build(this.genericAttributeFactory);
    graph.getNode(id, nodeType);
    this.thenGraphApproved(graph);
  }

  @Test
  public void itCanCreateGraphWithBranches() {
    var firstBranch = new GraphBuilder();
    firstBranch.addNode().setId(UniversallyUniqueIdentifier.randomUUID())
        .setType("first_branch_node");
    var secondBranch = firstBranch.createNewBranch();
    secondBranch.addEdge().setEdgeDirection(EdgeDirection.OUTGOING)
        .setId(UniversallyUniqueIdentifier.randomUUID()).setType("second_branch_edge");
    secondBranch.createNewBranch().addNode().setId(UniversallyUniqueIdentifier.randomUUID())
        .setType("third_branch_node");
    var graph = firstBranch.build(this.genericAttributeFactory);
    this.thenGraphApproved(graph);
  }

  @Test
  public void itCanCreateGraphWithMultipleBranches() {
    var firstBranch = new GraphBuilder();
    firstBranch.addNode().setId(UniversallyUniqueIdentifier.randomUUID())
        .setType("first_branch_node");
    var secondBranch = firstBranch.createNewBranch();
    secondBranch.addEdge().setEdgeDirection(EdgeDirection.OUTGOING)
        .setId(UniversallyUniqueIdentifier.randomUUID()).setType("second_branch_edge");
    secondBranch.addNode().setId(UniversallyUniqueIdentifier.randomUUID())
        .setType("second_branch_node");
    var thirdBranch = firstBranch.createNewBranch();
    thirdBranch.addEdge().setEdgeDirection(EdgeDirection.OUTGOING)
        .setId(UniversallyUniqueIdentifier.randomUUID()).setType("third_branch_edge");
    thirdBranch.addNode().setId(UniversallyUniqueIdentifier.randomUUID())
        .setType("third_branch_node");
    var graph = firstBranch.build(this.genericAttributeFactory);
    this.thenGraphApproved(graph);
  }

  @Test
  public void itCanCreateGraphWithElementsForRemoval() {
    var builder = new GraphBuilder();
    builder.addNodeForRemoval()
        .setType("node_type")
        .setId(UniversallyUniqueIdentifier.fromString("0d15149f-a0a8-42d1-a1bd-891355a637ae"));
    builder.addNodeForRemoval()
        .setType("another_type")
        .setId(UniversallyUniqueIdentifier.fromString("45e4088d-2ca1-410c-9a88-98be01d585a6"));
    var elements = builder.buildElementsForRemoval();
    elements.sort((e1, e2) -> e1.getGraphElementId().compareTo(e2.getGraphElementId()));
    this.thenObjectApproved(elements, this.getOptions());
  }

  private ObjectToJSonStringOptions getOptions() {
    return new ObjectToJSonStringOptions(
        ObjectToJSonStringOptions.RenderFeature.HIDE_IDS,
        ObjectToJSonStringOptions.RenderFeature.HIDE_CREATED_AT,
        ObjectToJSonStringOptions.RenderFeature.SORT_FIELDS
    );
  }
}
