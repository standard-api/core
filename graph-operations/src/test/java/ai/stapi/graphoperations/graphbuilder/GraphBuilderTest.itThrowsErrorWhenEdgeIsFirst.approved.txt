Graph can not have edge as first element.